# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [2]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Gustavo de Carvalho Azzola 76261
- Integrante 2:
- Integrante 3 (se houver):
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 60

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [4]:
url = "https://swapi.dev/api/people/60/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Gregar Typho",
    "height": "185",
    "mass": "85",
    "hair_color": "black",
    "skin_color": "dark",
    "eye_color": "brown",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/8/",
    "films": [
        "https://swapi.dev/api/films/5/"
    ],
    "species": [],
    "vehicles": [],
    "starships": [
        "https://swapi.dev/api/starships/39/"
    ],
    "created": "2014-12-20T11:10:10.381000Z",
    "edited": "2014-12-20T21:17:50.445000Z",
    "url": "https://swapi.dev/api/people/60/"
}


**Respostas do aluno — Personagem 60**

- Status code: 200
- Nome do personagem: Gregar Typhon
- Altura: 185
- Cor dos olhos: brown
- O JSON parece completo? Sim

### 1.2 Personagem 80

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [6]:
url = "https://swapi.dev/api/people/80/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Tarfful",
    "height": "234",
    "mass": "136",
    "hair_color": "brown",
    "skin_color": "brown",
    "eye_color": "blue",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/14/",
    "films": [
        "https://swapi.dev/api/films/6/"
    ],
    "species": [
        "https://swapi.dev/api/species/3/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T19:46:34.209000Z",
    "edited": "2014-12-20T21:17:50.491000Z",
    "url": "https://swapi.dev/api/people/80/"
}


**Respostas do aluno — Personagem 13**

- Status code: 200
- Nome do personagem: Tarfful
- Altura: 234
- Cor dos olhos: brown
- O JSON parece completo? sim

### 1.3 Planeta 100
**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [7]:
url = "https://swapi.dev/api/planets/100/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: fallback local
Status Code: 404
{
    "detail": "Not found"
}


**Respostas do aluno — Planeta 100**

- Status code: 404
- Nome do planeta: "detail": "Not found"
- Clima: "detail": "Not found"
- Terreno: "detail": "Not found"
- População: "detail": "Not found"
- O JSON parece completo? Não

### 1.4 Nave 3

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [9]:
url = "https://swapi.dev/api/starships/03/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Star Destroyer",
    "model": "Imperial I-class Star Destroyer",
    "manufacturer": "Kuat Drive Yards",
    "cost_in_credits": "150000000",
    "length": "1,600",
    "max_atmosphering_speed": "975",
    "crew": "47,060",
    "passengers": "n/a",
    "cargo_capacity": "36000000",
    "consumables": "2 years",
    "hyperdrive_rating": "2.0",
    "MGLT": "60",
    "starship_class": "Star Destroyer",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-10T15:08:19.848000Z",
    "edited": "2014-12-20T21:23:49.870000Z",
    "url": "https://swapi.dev/api/starships/3/"
}


**Respostas do aluno — Nave 3**

- Status code: 200
- Nome da nave: Star Destroyer
- Modelo:  Imperial I-class Star Destroyer
- Fabricante: Kuat Drive Yards
- Velocidade máxima em atmosfera: 975
- O JSON parece completo? Sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [10]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Obi-Wan Kenobi
Altura: 182
Cor dos olhos: blue-gray
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome?
dados["name"]
2. Qual campo foi usado para mostrar a altura?
dados["height"]
3. Se você fosse QA, quais desses campos validaria primeiro?
Eu validaria primeiro o campo height (Altura).

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [11]:
url = "https://swapi.dev/api/planets/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Alderaan
Clima: temperate
Terreno: grasslands, mountains
População: 2000000000


**Perguntas**
1. O campo população parece numérico? Não pois normalmente API retornam esses valores em aspas como uma string
2. O clima foi retornado corretamente? Sim
3. Se você fosse QA, o que validaria nesse planeta? Validação de Tipo e Conversão: Verificar se o campo population pode ser convertido para número sem erros.

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [12]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? Esta validando se a solicitação esta sendo recebida e executada com sucesso
2. O que a validação 2 verifica?
Verifica se a API retorna um campo "name"
3. O que a validação 3 verifica?
Verifica se o a API esta retornando um nome
4. O que a validação 4 verifica?
Verifica se retorna um campo com height

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [13]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


Resposta
1. Validação: response.status_code == 200
Por que faz sentido: É o primeiro portão. Se o status não for 200 (OK), significa que a API falhou (erro 500), o endereço está errado (erro 404) ou você não tem permissão (erro 403).

Visão de QA: Não adianta testar o nome da nave se a porta do servidor está fechada. É a validação de disponibilidade.

2. Validação: "name" in dados e "model" in dados
Por que faz sentido: Garante que a estrutura do JSON (o "contrato") não mudou.

Visão de QA: Imagine que o time de desenvolvimento mude a chave de name para ship_name. Se o seu front-end espera name, o site vai quebrar ou exibir "Undefined". Validar a existência da chave previne que quebras na interface cheguem ao usuário final.

3. Validação: dados["name"] != ""
Por que faz sentido: Verifica se o campo está populado. Um campo pode existir na estrutura, mas vir vazio ou nulo.

Visão de QA: De nada serve o campo existir se ele não traz a informação. Para o usuário, ver um card de nave sem nome é um erro de experiência (UX) tão grave quanto um erro de sistema.

## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [14]:
url = "https://swapi.dev/api/people/13/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


respotas

Diferente do comando print, o comando assert é uma ferramenta fundamental de Testes Automatizados. Ele funciona como uma "sentinela": se a condição for verdadeira, o código segue em frente; se for falsa, o programa interrompe a execução na hora e exibe a mensagem de erro que você escreveu.

O que cada assert valida?
assert response.status_code == 200

O que valida: A comunicação com o servidor.

Por que é vital: Garante que a API está "viva" e respondeu com sucesso. Se o servidor cair ou o link estiver errado, o teste para aqui mesmo, evitando que você tente analisar dados que não existem.

assert "name" in dados

O que valida: A estrutura do contrato (Schema).

Por que é vital: Confirma que a API ainda entrega a chave name. Se os desenvolvedores da API mudarem o nome do campo para full_name, esse teste falha, avisando que o sistema que consome esses dados vai quebrar.

assert dados["name"] != ""

O que valida: A qualidade do conteúdo.

Por que é vital: Garante que o campo não está vindo como uma string vazia. De nada adianta a estrutura estar certa se a informação principal (o nome do personagem) estiver faltando.

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [15]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


resposta

Detalhamento dos Campos Validados
response.status_code == 200

Campo: Cabeçalho de Resposta HTTP (Status Code).

Objetivo: Validar se a requisição foi bem-sucedida. Se o servidor retornasse 404 (Não encontrado) ou 500 (Erro interno), o teste travaria aqui.

"manufacturer" in dados

Campo: manufacturer (Fabricante).

Objetivo: Validar a existência da chave no JSON. Garante que o contrato de dados está sendo respeitado e que o campo "fabricante" não foi removido da API.

dados["model"] != ""

Campo: model (Modelo).

Objetivo: Validar se o conteúdo não está vazio. Confirma que, além de existir a chave, ela traz uma informação útil (ex: "YT-1300 light freighter").

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [ ]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado?
404
2. Esse comportamento está correto?
Sim, está perfeitamente correto.
Se você pede algo que não existe, o sistema não pode retornar "sucesso" (200) nem uma página em branco sem explicação.
3. Por que um QA deve testar cenários negativos?
Evitar o "Crash" do Sistema
Garantir Mensagens de Erro Claras
Segurança e Robustez: Testar o que acontece quando enviamos dados inválidos

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [16]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/people/4/")

Origem dos dados: online
Status Code: 200
name: Darth Vader
height: 202
mass: 136
hair_color: none
skin_color: white
eye_color: yellow
birth_year: 41.9BBY
gender: male
homeworld: https://swapi.dev/api/planets/1/
films: ['https://swapi.dev/api/films/1/', 'https://swapi.dev/api/films/2/', 'https://swapi.dev/api/films/3/', 'https://swapi.dev/api/films/6/']
species: []
vehicles: []
starships: ['https://swapi.dev/api/starships/13/']
created: 2014-12-10T15:18:20.704000Z
edited: 2014-12-20T21:17:50.313000Z
url: https://swapi.dev/api/people/4/


**Respostas finais**
1. Qual endpoint você testou no desafio?
"https://swapi.dev/api/people/4/"
2. Qual foi o status code?
200
3. Qual campo chamou mais a sua atenção?
O campo starships (naves) é muito interessante. Note que ele contém apenas uma URL: 'https://swapi.dev/api/starships/13/'.

Diferente de outros personagens que pilotam várias naves, o Vader está vinculado especificamente ao seu TIE Advanced. Para um desenvolvedor ou QA, esse campo é um "ponteiro"; ele não te dá o nome da nave, mas te obriga a fazer uma segunda requisição à API para descobrir que nave é aquela.

4. Se você fosse QA, quais 3 validações faria nesse recurso?
Validação de "Ponteiros" de Lista (starships e films)
O que testar: Verificar se os links contidos nas listas são válidos.

Por que: Como QA, eu faria um teste automatizado para clicar em cada link da lista de filmes e naves. Se algum deles retornar 404 Not Found, a integridade do banco de dados está comprometida e o usuário encontrará um link quebrado no site.

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

 Status Code (Sucesso e Erro): Garantir que a API responda 200 OK para dados
existentes e 404 Not Found para IDs inexistentes. É a base da confiabilidade do sistema.

 Integridade do Contrato (Schema): Validar se chaves essenciais como name, model ou starships estão presentes no JSON. Se o nome da chave mudar no servidor, o sistema que consome os dados para de funcionar.

 Consistência de Tipos e Conteúdo: Verificar se campos que deveriam ser números (como height ou mass) não estão vazios ou com textos inesperados (como "desconhecido"). Isso evita erros de cálculo e quebras na interface do usuário.

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
